plugins/scheme-tpm-enacttrust/main.go

```diff
@@ -33,15 +55,108 @@ func (s Scheme) SynthKeysFromTrustAnchor(tenantID string, ta *common.TrustAnchor
-	return nil, errors.New("TODO")
+func (s Scheme) ExtractEvidence(
+	token *common.AttestationToken,
+	trustAnchor string,
+) (*common.ExtractedEvidence, error) {
+	if token.Format != common.AttestationFormat_TPM_ENACTTRUST {
+		return nil, fmt.Errorf("wrong format: expect %q, but found %q",
+			common.AttestationFormat_TPM_ENACTTRUST.String(),
+			token.Format.String(),
+		)
+	}
+
+	var decoded Token
+
+	if err := decoded.Decode(token.Data); err != nil {
+		return nil, fmt.Errorf("could not decode token: %v", err)
+	}
+
+	pubKey, err := parseKey(trustAnchor)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse trust anchor: %v", err)
+	}
+
+	if err = decoded.VerifySignature(pubKey); err != nil {
+		return nil, fmt.Errorf("could not verify signature: %v", err)
+	}
+
+	if decoded.AttestationData.Type != tpm2.TagAttestQuote {
+		return nil, fmt.Errorf("wrong TPMS_ATTEST type: expected %d, but got %d",
+			tpm2.TagAttestQuote, decoded.AttestationData.Type)
+	}
+
+	var pcrs []int64
+	for _, pcr := range decoded.AttestationData.AttestedQuoteInfo.PCRSelection.PCRs {
+		pcrs = append(pcrs, int64(pcr))
+	}
+
+	evidence := common.NewExtractedEvidence()
+	evidence.Evidence["pcr-selection"] = pcrs
+	evidence.Evidence["hash-algorithm"] = int64(decoded.AttestationData.AttestedQuoteInfo.PCRSelection.Hash)
+	evidence.Evidence["pcr-digest"] = []byte(decoded.AttestationData.AttestedQuoteInfo.PCRDigest)
+
+	nodeID, err := uuid.FromBytes(decoded.AttestationData.ExtraData)
+	if err != nil {
+		return nil, fmt.Errorf("could not decode node-id: %v", err)
+	}
+	evidence.SoftwareID = fmt.Sprintf("tpm-enacttrust://%d/%v", token.TenantId, nodeID.String())
+
+	return evidence, nil
 }
 
-func (s Scheme) GetAttestation(ec *common.EvidenceContext, endorsements []string) (*common.Attestation, error) {
-	return nil, errors.New("TODO")
+func (s Scheme) GetAttestation(
+	ec *common.EvidenceContext,
+	endorsementStrings []string,
+) (*common.Attestation, error) {
+
+	attestation := common.NewAttestation(ec)
+
+	attestation.Result.TrustVector.SoftwareIntegrity = common.AR_Status_FAILURE
+	attestation.Result.Status = common.AR_Status_FAILURE
+
+	digestValue, ok := ec.Evidence.AsMap()["pcr-digest"]
+	if !ok {
+		return attestation, fmt.Errorf("evidence does not contain \"pcr-digest\" entry")
+	}
+
+	evidenceDigest, ok := digestValue.(string)
+	if !ok {
+		return attestation, fmt.Errorf("wrong type value \"pcr-digest\" entry; expected string but found %T", digestValue)
+	}
+
+	var endorsements Endorsements
+	if err := endorsements.Populate(endorsementStrings); err != nil {
+		return attestation, fmt.Errorf("incorrect endorsements number: expected exactly 1, but found %v", endorsements)
+	}
+
+	if endorsements.Digest == evidenceDigest {
+		attestation.Result.TrustVector.SoftwareIntegrity = common.AR_Status_SUCCESS
+		attestation.Result.Status = common.AR_Status_SUCCESS
+	}
+
+	return attestation, nil
```

Comment:
Perhaps the functions on line 163 and 180 have been copied here from PSA specific plugin.

Could you please factor them to more common place or atleast correct the comment above each of this 9in the first iteration) as it is quite misleading the way they are ?

Reply:
These were pre-existing as part of the provisioning paths. I'm uncomfortable touching them as part of this changeset (this is not part of this pull request), but will update the comments with the correct scheme name.